optaplanner-core/src/test/java/org/optaplanner/core/api/score/stream/JoinersTest.java

```diff
@@ -300,9 +308,163 @@ public void greaterThanOrEqualPenta() {
+                    BigDecimal.TEN)).isTrue();
+            softly.assertThat(joiner.matches(BigInteger.TEN, BigInteger.ZERO, BigInteger.ZERO, BigInteger.ZERO,
+                    BigDecimal.ONE)).isTrue();
+            softly.assertThat(joiner.matches(BigInteger.ZERO, BigInteger.ZERO, BigInteger.ZERO, BigInteger.ZERO,
+                    BigDecimal.ONE)).isFalse();
+        });
+    }
+
+    private static final class Interval<A, B> {
+        final A start;
+        final A end;
+        final Function<A, B> map;
+
+        public Interval(A start, A end, Function<A, B> map) {
+            this.start = start;
+            this.end = end;
+            this.map = map;
+        }
+
+        public B getStart() {
+            return map.apply(start);
+        }
+
+        public B getEnd() {
+            return map.apply(end);
+        }
+
+        @Override
+        public String toString() {
+            return "(" + start.toString() + ", " + end.toString() + ")";
+        }
+
+        public static <A, B, C> IntervalFactory<A, B, C> createFactory(Function<C, A> constructorMap, Function<A, B> map) {
+            return new IntervalFactory<A, B, C>(constructorMap, map);
+        }
+
+        private final static class IntervalFactory<A, B, C> {
+            final Function<C, A> constructorMap;
+            final Function<A, B> map;
+
+            public IntervalFactory(Function<C, A> constructorMap, Function<A, B> map) {
+                this.constructorMap = constructorMap;
+                this.map = map;
+            }
+
+            public Interval<A, B> of(C start, C end) {
+                return new Interval<A, B>(constructorMap.apply(start), constructorMap.apply(end), map);
+            }
+        }
+    }
+
+    private final Interval.IntervalFactory<BigInteger, Long, Long> bigIntInterval =
+            Interval.createFactory(BigInteger::valueOf, BigInteger::longValue);
+
+    private final Interval.IntervalFactory<BigDecimal, Long, Long> bigDecimalInterval =
+            Interval.createFactory(BigDecimal::valueOf, BigDecimal::longValue);
+
+    @Test
+    public void overlapsBi() {
+        Function<Interval<BigInteger, Long>, Long> leftStartMapping = Interval::getStart;
+        Function<Interval<BigInteger, Long>, Long> leftEndMapping = Interval::getEnd;
+        Function<Interval<BigDecimal, Long>, Long> rightStartMapping = Interval::getStart;
+        Function<Interval<BigDecimal, Long>, Long> rightEndMapping = Interval::getEnd;
+
+        AbstractBiJoiner<Interval<BigInteger, Long>, Interval<BigDecimal, Long>> joiner =
+                (AbstractBiJoiner<Interval<BigInteger, Long>, Interval<BigDecimal, Long>>) Joiners.overlaps(leftStartMapping,
+                        leftEndMapping, rightStartMapping, rightEndMapping);
+
+        assertSoftly(softly -> {
+            // Case A = B
+            softly.assertThat(joiner.matches(bigIntInterval.of(1L, 5L),
```

Comment:
Now that the Joiners are actually relatively complex, I think that the comment before these assertions (`Case A = B`) should instead be turned into `.as(...)` on the assertion itself. The assertion messages will become so much better for it.

Reply:
(Which wouldn't be necessary if each of the cases had their own properly named test methods, as I suggest above. Either is fine with me.)